### PR TITLE
Private binary caches support

### LINKED ISSUE #158: Don't exit with non-zero on NixOS
`cachix use <name>` should exit with 0.

### PR DESCRIPTION
**Pull request by domenkozar**
Fixes #158 #160 #166 #91 #24 #170 #168 #71

### CODE REVIEW

File: cachix/src/Cachix/Client/InstallationMode.hs [hunk 1]
```diff
@@ -57,50 +48,107 @@ getInstallationMode NixEnv{..}
...
+
+    imports = [ ./cachix.nix ];
+
+Then run:
+
+    $ nixos-rebuild switch
+    |]
+
+    glueModule :: Text
+    glueModule = [i|
```
Comment (roberth) [inline cachix/src/Cachix/Client/InstallationMode.hs]: Warn that this file will be overwritten by the cachix use command.

File: cachix/src/Cachix/Client/NetRc.hs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import           System.Directory     ( doesFileExist )
+import           Servant.Auth.Client  (getToken)
+
+import qualified Cachix.Api as Api
+import           Cachix.Api.Error     (escalateAs)
+import           Cachix.Client.Config (Config, authToken)
+import           Cachix.Client.Exception (CachixException(NetRcParseError))
+
+
+{- Add a list of binary caches to netrc under `filename`.
```
Comment (roberth) [inline cachix/src/Cachix/Client/NetRc.hs]: Why not haddock

File: cachix/test/NetRcSpec.hs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  , githubUsername = "foobar2"
+  }
+
+config :: Config
+config = Config
+  { authToken = "token123"
+  , binaryCaches = []
+  }
+
+-- TODO: poor man's golden tests, hspec doesn't have a lib
```
Comment (roberth) [inline cachix/test/NetRcSpec.hs]:
```suggestion

```
I don't think anyone cares. Works fine, right?

File: cachix/test/NetRcSpec.hs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    expected <- readFile output
+    real `shouldBe` expected
+  where
+    getInput = getDataFileName $ toS goldenName <> ".input"
+    getOutput = getDataFileName $ toS goldenName <> ".output"
+
+spec :: Spec
+spec =
+  describe "add" $ do
+    -- TODO: not easy to test this with temp files as they are *created*
```
Comment (roberth) [inline cachix/test/NetRcSpec.hs]: Use a temp directory: `withSystemTempDirectory`

File: cachix/src/Cachix/Client/InstallationMode.hs [hunk 5]
```diff
@@ -57,50 +48,107 @@ getInstallationMode NixEnv{..}
...
 
+nixosBinaryCache :: Maybe Config -> Maybe Text -> Api.BinaryCache -> UseOptions -> IO ()
+nixosBinaryCache maybeConfig maybeUser bc@Api.BinaryCache{..} UseOptions { useNixOSFolder=baseDirectory } = do
+  createDirectoryIfMissing True $ toS toplevel
+  writeFile (toS glueModuleFile) glueModule
+  writeFile (toS cacheModuleFile) cacheModule
+  unless isPublic $ addPrivateBinaryCacheNetRC maybeConfig bc NixConf.Global
+  putText instructions
+  where
+    configurationNix :: Text
```
[the remark this answers is not in the document]
Comment (domenkozar) [reply cachix/src/Cachix/Client/InstallationMode.hs]: I don't like that syntax, it's too dense and easy to miss function signature.

File: cachix/src/Cachix/Client/Commands.hs [hunk 6]
```diff
@@ -114,15 +122,22 @@ To instruct Nix to use the binary cache:
...
+  maybe (Token "") authToken (config env)
 
-use :: Env -> Text -> Bool -> IO ()
-use env name shouldEchoNixOS = do
+use :: Env -> Text -> UseOptions -> IO ()
+use env name useOptions = do
   -- 1. get cache public key
-  res <- (`runClientM` clientenv env) $ Api.get (cachixBCClient name)
+  res <- (`runClientM` clientenv env) $ Api.get (cachixBCClient name) (envToToken env)
   case res of
```
[the remark this answers is not in the document]
Comment (domenkozar) [reply cachix/src/Cachix/Client/Commands.hs]: It's not the same error logic, don't think there's a lot of DRYness to be achieved, maybe can make the 401 case DRY though.

File: cachix/test/NetRcSpec.hs (hunk 3 above)
Comment (domenkozar) [reply cachix/test/NetRcSpec.hs]: Going to add a comment to use https://github.com/stackbuilders/hspec-golden one day

Comment (roberth) [reply cachix/test/NetRcSpec.hs]: Gold plating? ;)

File: cachix/test/NetRcSpec.hs (hunk 4 above)
Comment (domenkozar) [reply cachix/test/NetRcSpec.hs]: Yeah needs to be refactored, not this time.

File: cachix/test/NetRcSpec.hs (hunk 3 above)
Comment (domenkozar) [reply cachix/test/NetRcSpec.hs]: It didn't exist when I wrote this code, but I'd rather have smaller codebase footprint. Very low impact, low priority :)

**Comment by domenkozar** [comment]
I've addressed most review comments, added my reasoning for those that I didn't :)

File: cachix/src/Cachix/Client/NetRc.hs [hunk 7]
```diff
@@ -16,9 +16,9 @@ import           Cachix.Client.Config (Config, authToken)
 import           Cachix.Client.Exception (CachixException(NetRcParseError))
 
 
-{- Add a list of binary caches to netrc under `filename`.
-   Makes sure there are no duplicate entries (using domain as a key).
-   If file under filename doesn't exist it's created.
+{- | Add a list of binary caches to netrc under `filename`.
+   | Makes sure there are no duplicate entries (using domain as a key).
```
Comment (roberth) [inline cachix/src/Cachix/Client/NetRc.hs]: Doesn't that introduce `|`s in the output?

I think we should standardize on
```
-- |
--
```

Similar problem here: https://hackage.haskell.org/package/cachix-0.1.1/docs/Cachix-Client-URI.html

File: cachix/src/Cachix/Client/Commands.hs (hunk 6 above)
[the remark this answers is not in the document]
Comment (roberth) [reply cachix/src/Cachix/Client/Commands.hs]: Ok, so improveError can be a lambda instead

[the remark this answers is not in the document]
Comment (domenkozar) [reply cachix/src/Cachix/Client/Commands.hs]: What about the implementation be? I don't think it easily generalizes.

[the remark this answers is not in the document]
Comment (roberth) [reply cachix/src/Cachix/Client/Commands.hs]: It's just refactoring, that's all. Not important.

**Comment by domenkozar** [comment]
@roberth ready for re-review, added LTS-13 support.

**Pull request merged**